admin/commands/storage/read_blocks_test.go

```diff
@@ -0,0 +1,253 @@
+)
+
+type ReadBlocksSuite struct {
+	suite.Suite
+
+	command commands.AdminCommand
+	state   *protocolmock.State
+	blocks  *storagemock.Blocks
+
+	final     *flow.Block
+	sealed    *flow.Block
+	allBlocks []*flow.Block
+}
+
+func TestReadBlocks(t *testing.T) {
+	t.Parallel()
+	suite.Run(t, new(ReadBlocksSuite))
+}
+
+func createSnapshot(head *flow.Header) protocol.Snapshot {
+	snapshot := &protocolmock.Snapshot{}
+	snapshot.On("Head").Return(
+		func() *flow.Header {
+			return head
+		},
+		nil,
+	)
+	return snapshot
+}
+
+func (suite *ReadBlocksSuite) SetupTest() {
+	suite.state = new(protocolmock.State)
+	suite.blocks = new(storagemock.Blocks)
+
+	var blocks []*flow.Block
+
+	genesis := unittest.GenesisFixture()
+	blocks = append(blocks, genesis)
+	sealed := unittest.BlockWithParentFixture(genesis.Header)
+	blocks = append(blocks, sealed)
+	final := unittest.BlockWithParentFixture(sealed.Header)
+	blocks = append(blocks, final)
+	final = unittest.BlockWithParentFixture(final.Header)
+	blocks = append(blocks, final)
+	final = unittest.BlockWithParentFixture(final.Header)
+	blocks = append(blocks, final)
+
+	suite.allBlocks = blocks
+	suite.sealed = sealed
+	suite.final = final
+
+	suite.state.On("Final").Return(createSnapshot(final.Header))
+	suite.state.On("Sealed").Return(createSnapshot(sealed.Header))
+	suite.state.On("AtBlockID", mock.Anything).Return(
+		func(blockID flow.Identifier) protocol.Snapshot {
+			for _, block := range blocks {
+				if block.ID() == blockID {
+					return createSnapshot(block.Header)
+				}
+			}
+			return invalid.NewSnapshot(fmt.Errorf("invalid block ID: %v", blockID))
+		},
+	)
+	suite.state.On("AtHeight", mock.Anything).Return(
+		func(height uint64) protocol.Snapshot {
+			if int(height) < len(blocks) {
+				block := blocks[height]
+				return createSnapshot(block.Header)
+			}
+			return invalid.NewSnapshot(fmt.Errorf("invalid height: %v", height))
+		},
+	)
+
+	suite.blocks.On("ByID", mock.Anything).Return(
+		func(blockID flow.Identifier) *flow.Block {
+			for _, block := range blocks {
+				if block.ID() == blockID {
+					return block
+				}
+			}
+			return nil
+		},
+		func(blockID flow.Identifier) error {
+			for _, block := range blocks {
+				if block.ID() == blockID {
+					return nil
+				}
+			}
+			return fmt.Errorf("block %#v not found", blockID)
+		},
+	)
+
+	suite.command = NewReadBlocksCommand(suite.state, suite.blocks)
+}
+
+func (suite *ReadBlocksSuite) TestValidateInvalidFormat() {
+	assert.Error(suite.T(), suite.command.Validator(&admin.CommandRequest{
+		Data: true,
+	}))
+	assert.Error(suite.T(), suite.command.Validator(&admin.CommandRequest{
+		Data: 420,
+	}))
+	assert.Error(suite.T(), suite.command.Validator(&admin.CommandRequest{
+		Data: "foo",
+	}))
+	assert.Error(suite.T(), suite.command.Validator(&admin.CommandRequest{
+		Data: map[string]interface{}{
+			"glock": 123,
+		},
+	}))
+}
+
+func (suite *ReadBlocksSuite) TestValidateInvalidBlock() {
```

Comment:
Followup to suggestion above, should we return sentinel errors for these invalid data inputs ?